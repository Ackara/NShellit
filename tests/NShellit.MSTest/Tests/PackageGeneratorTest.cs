using Acklann.NShellit.Extensions;
using Acklann.NShellit.Fakes;
using Acklann.NShellit.Generators;
using ApprovalTests;
using ApprovalTests.Namers;
using Microsoft.VisualStudio.TestTools.UnitTesting;
using Shouldly;
using System;
using System.Collections.Generic;
using System.IO;
using System.Linq;

namespace Acklann.NShellit.Tests
{
    [TestClass]
    public class PackageGeneratorTest
    {
        [TestMethod]
        public void PSModuleGenerator_should_create_a_publishable_powershell_module()
        {
            var sut = new PSModuleGenerator();
            RunPackageTest(sut, filesToExamine: () => (from f in Directory.EnumerateFiles(sut.PackageDirectory, "*", SearchOption.AllDirectories)
                                                       where Path.GetExtension(f) == ".ps1" /*|| Path.GetExtension(f) == ".psd1"*/
                                                       select f));
        }

        //[TestMethod]
        public void NodeModuleGenerator_should_create_a_publishable_powershell_module()
        {
            IShellWrapper sut = null;
            RunPackageTest(sut, filesToExamine: () => (from f in Directory.EnumerateFiles(sut.PackageDirectory, "*", SearchOption.AllDirectories)
                                                       where Path.GetExtension(f) == ".js"
                                                       select f));
        }

        private static IEnumerable<CommandInfo> GetCommandList()
        {
            var one = new CommandInfo<NonDecoratedCommand>("one") { Cmdlet = "Write-Foo" };
            one.SetDescription("This is a good command.")
                .WithParameter(x => x.NumericValue)
                    .SetDescription("An id value.")
                    .Required()
                .WithParameter(x => x.PrecisionValue, "coordinate")
                    .SetDescription("An number value.")
                    .SetAlias("point")
                    .SetDefault(0)
                .WithParameter(x => x.CharValue)
                    .SetDescription("A char value.")
                    .SetAlias("c")
                    .SetDefault('c')
                .WithParameter(x => x.Collection)
                    .SetDescription("An array")
                    .SetAlias("l", "list")
                    .SetDefault(new string[] { "a", "b", "c" })
                .WithParameter(x => x.DateValue)
                    .SetDescription("A date value.")
                    .SetAlias("d", "date")
                    .SetDefault(DateTime.Now)
                .WithParameter(x => x.EnumValue)
                    .SetDescription("A fake value.")
                    .SetAlias("e")
                    .SetDefault(FakeEnum.North);

            one.AddLinks("Write-Host");
            one.AddExample("Write-Foo", "This does something.");
            one.AddExample("Write-Foo", "This does something.", true);
            
            var two = new CommandInfo("two");
            two.SetDescription("This is a synopsis.")
                .WithParameter<string>("Id")
                .SetDescription("An id value.")
                .Required()
                .WithParameter("Contact", typeof(string), "phone")
                .SetDescription("A phone number")
                .SetAlias("tel");

            return new CommandInfo[] { one, two };
        }

        private static void RunPackageTest(IShellWrapper sut, Func<IEnumerable<string>> filesToExamine)
        {
            //  Act
            sut.GeneratePackage(GetCommandList(), typeof(PackageGeneratorTest).Assembly.Location);
            var package = new DirectoryInfo(sut.PackageDirectory);
            var fileList = (from f in package.GetFiles("*", SearchOption.AllDirectories)
                            select f.FullName.Replace(package.FullName, string.Empty));

            // Assert
            package.Exists.ShouldBeTrue();

            using (ApprovalResults.ForScenario("fileList"))
            {
                Approvals.VerifyAll(fileList, "PS");
            }

            foreach (string file in filesToExamine())
            {
                using (ApprovalResults.ForScenario(Path.GetFileNameWithoutExtension(file)))
                {
                    Approvals.Verify(File.ReadAllText(file));
                }
            }
        }
    }
}