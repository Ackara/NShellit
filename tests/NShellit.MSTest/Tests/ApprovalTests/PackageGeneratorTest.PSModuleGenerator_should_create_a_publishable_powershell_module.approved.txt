PS[0] = \Poshsielaye.psd1
PS[1] = \bin\Acklann.Poshley.dll
PS[2] = \bin\Acklann.Poshley.MSTest.dll
PS[3] = \bin\Acklann.Poshley.MSTest.pdb
PS[4] = \bin\Acklann.Poshley.pdb
PS[5] = \bin\Acklann.Poshsielaye.dll
PS[6] = \bin\Acklann.Poshsielaye.MSTest.dll
PS[7] = \bin\Acklann.Poshsielaye.MSTest.dll.config
PS[8] = \bin\Acklann.Poshsielaye.MSTest.pdb
PS[9] = \bin\Acklann.Poshsielaye.pdb
PS[10] = \bin\ApprovalTests.dll
PS[11] = \bin\ApprovalTests.xml
PS[12] = \bin\ApprovalUtilities.dll
PS[13] = \bin\ApprovalUtilities.Net45.dll
PS[14] = \bin\ApprovalUtilities.Net45.xml
PS[15] = \bin\ApprovalUtilities.xml
PS[16] = \bin\Microsoft.VisualStudio.TestPlatform.MSTest.TestAdapter.dll
PS[17] = \bin\Microsoft.VisualStudio.TestPlatform.MSTestAdapter.PlatformServices.dll
PS[18] = \bin\Microsoft.VisualStudio.TestPlatform.MSTestAdapter.PlatformServices.Interface.dll
PS[19] = \bin\Microsoft.VisualStudio.TestPlatform.TestFramework.dll
PS[20] = \bin\Microsoft.VisualStudio.TestPlatform.TestFramework.Extensions.dll
PS[21] = \bin\Microsoft.VisualStudio.TestPlatform.TestFramework.Extensions.xml
PS[22] = \bin\Microsoft.VisualStudio.TestPlatform.TestFramework.xml
PS[23] = \bin\Microsoft.Win32.Primitives.dll
PS[24] = \bin\netstandard.dll
PS[25] = \bin\Newtonsoft.Json.dll
PS[26] = \bin\Newtonsoft.Json.xml
PS[27] = \bin\Shouldly.dll
PS[28] = \bin\System.AppContext.dll
PS[29] = \bin\System.Collections.Concurrent.dll
PS[30] = \bin\System.Collections.dll
PS[31] = \bin\System.Collections.NonGeneric.dll
PS[32] = \bin\System.Collections.Specialized.dll
PS[33] = \bin\System.ComponentModel.dll
PS[34] = \bin\System.ComponentModel.EventBasedAsync.dll
PS[35] = \bin\System.ComponentModel.Primitives.dll
PS[36] = \bin\System.ComponentModel.TypeConverter.dll
PS[37] = \bin\System.Console.dll
PS[38] = \bin\System.Data.Common.dll
PS[39] = \bin\System.Diagnostics.Contracts.dll
PS[40] = \bin\System.Diagnostics.Debug.dll
PS[41] = \bin\System.Diagnostics.FileVersionInfo.dll
PS[42] = \bin\System.Diagnostics.Process.dll
PS[43] = \bin\System.Diagnostics.StackTrace.dll
PS[44] = \bin\System.Diagnostics.TextWriterTraceListener.dll
PS[45] = \bin\System.Diagnostics.Tools.dll
PS[46] = \bin\System.Diagnostics.TraceSource.dll
PS[47] = \bin\System.Diagnostics.Tracing.dll
PS[48] = \bin\System.Drawing.Primitives.dll
PS[49] = \bin\System.Dynamic.Runtime.dll
PS[50] = \bin\System.Globalization.Calendars.dll
PS[51] = \bin\System.Globalization.dll
PS[52] = \bin\System.Globalization.Extensions.dll
PS[53] = \bin\System.IO.Compression.dll
PS[54] = \bin\System.IO.Compression.ZipFile.dll
PS[55] = \bin\System.IO.dll
PS[56] = \bin\System.IO.FileSystem.dll
PS[57] = \bin\System.IO.FileSystem.DriveInfo.dll
PS[58] = \bin\System.IO.FileSystem.Primitives.dll
PS[59] = \bin\System.IO.FileSystem.Watcher.dll
PS[60] = \bin\System.IO.IsolatedStorage.dll
PS[61] = \bin\System.IO.MemoryMappedFiles.dll
PS[62] = \bin\System.IO.Pipes.dll
PS[63] = \bin\System.IO.UnmanagedMemoryStream.dll
PS[64] = \bin\System.Linq.dll
PS[65] = \bin\System.Linq.Expressions.dll
PS[66] = \bin\System.Linq.Parallel.dll
PS[67] = \bin\System.Linq.Queryable.dll
PS[68] = \bin\System.Net.Http.dll
PS[69] = \bin\System.Net.NameResolution.dll
PS[70] = \bin\System.Net.NetworkInformation.dll
PS[71] = \bin\System.Net.Ping.dll
PS[72] = \bin\System.Net.Primitives.dll
PS[73] = \bin\System.Net.Requests.dll
PS[74] = \bin\System.Net.Security.dll
PS[75] = \bin\System.Net.Sockets.dll
PS[76] = \bin\System.Net.WebHeaderCollection.dll
PS[77] = \bin\System.Net.WebSockets.Client.dll
PS[78] = \bin\System.Net.WebSockets.dll
PS[79] = \bin\System.ObjectModel.dll
PS[80] = \bin\System.Reflection.dll
PS[81] = \bin\System.Reflection.Extensions.dll
PS[82] = \bin\System.Reflection.Primitives.dll
PS[83] = \bin\System.Resources.Reader.dll
PS[84] = \bin\System.Resources.ResourceManager.dll
PS[85] = \bin\System.Resources.Writer.dll
PS[86] = \bin\System.Runtime.CompilerServices.VisualC.dll
PS[87] = \bin\System.Runtime.dll
PS[88] = \bin\System.Runtime.Extensions.dll
PS[89] = \bin\System.Runtime.Handles.dll
PS[90] = \bin\System.Runtime.InteropServices.dll
PS[91] = \bin\System.Runtime.InteropServices.RuntimeInformation.dll
PS[92] = \bin\System.Runtime.Numerics.dll
PS[93] = \bin\System.Runtime.Serialization.Formatters.dll
PS[94] = \bin\System.Runtime.Serialization.Json.dll
PS[95] = \bin\System.Runtime.Serialization.Primitives.dll
PS[96] = \bin\System.Runtime.Serialization.Xml.dll
PS[97] = \bin\System.Security.Claims.dll
PS[98] = \bin\System.Security.Cryptography.Algorithms.dll
PS[99] = \bin\System.Security.Cryptography.Csp.dll
PS[100] = \bin\System.Security.Cryptography.Encoding.dll
PS[101] = \bin\System.Security.Cryptography.Primitives.dll
PS[102] = \bin\System.Security.Cryptography.X509Certificates.dll
PS[103] = \bin\System.Security.Principal.dll
PS[104] = \bin\System.Security.SecureString.dll
PS[105] = \bin\System.Text.Encoding.dll
PS[106] = \bin\System.Text.Encoding.Extensions.dll
PS[107] = \bin\System.Text.RegularExpressions.dll
PS[108] = \bin\System.Threading.dll
PS[109] = \bin\System.Threading.Overlapped.dll
PS[110] = \bin\System.Threading.Tasks.dll
PS[111] = \bin\System.Threading.Tasks.Parallel.dll
PS[112] = \bin\System.Threading.Thread.dll
PS[113] = \bin\System.Threading.ThreadPool.dll
PS[114] = \bin\System.Threading.Timer.dll
PS[115] = \bin\System.ValueTuple.dll
PS[116] = \bin\System.Xml.ReaderWriter.dll
PS[117] = \bin\System.Xml.XDocument.dll
PS[118] = \bin\System.Xml.XmlDocument.dll
PS[119] = \bin\System.Xml.XmlSerializer.dll
PS[120] = \bin\System.Xml.XPath.dll
PS[121] = \bin\System.Xml.XPath.XDocument.dll
PS[122] = \bin\Telerik.JustMock.dll
PS[123] = \bin\Telerik.JustMock.xml
PS[124] = \bin\TestData\packageMe.txt
